pkg/asset/manifests/ingress.go

```diff
@@ -0,0 +1,137 @@
+package manifests
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+
+	configv1 "github.com/openshift/api/config/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	ingCrdFilename = filepath.Join(manifestDir, "cluster-ingress-01-crd.yml")
+	ingCfgFilename = filepath.Join(manifestDir, "cluster-ingress-02-config.yml")
+)
+
+const (
+	ingressConfigCRD = `
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: ingresses.config.openshift.io
+spec:
+  group: config.openshift.io
+  names:
+    kind: Ingress
+    listKind: IngressList
+    plural: ingresses
+    singular: ingress
+  scope: Cluster
+  versions:
+    - name: v1
+      served: true
+      storage: true
+`
+)
+
+// Ingress generates the cluster-ingress-*.yml files.
+type Ingress struct {
+	config   *configv1.Ingress
+	FileList []*asset.File
+}
+
+var _ asset.WritableAsset = (*Ingress)(nil)
+
+// Name returns a human friendly name for the asset.
+func (*Ingress) Name() string {
+	return "Ingress Config"
+}
+
+// Dependencies returns all of the dependencies directly needed to generate
+// the asset.
+func (*Ingress) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		&installconfig.InstallConfig{},
+	}
+}
+
+// Generate generates the ingress config and its CRD.
+func (ing *Ingress) Generate(dependencies asset.Parents) error {
+	installConfig := &installconfig.InstallConfig{}
+	dependencies.Get(installConfig)
+
+	ing.config = &configv1.Ingress{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: configv1.SchemeGroupVersion.String(),
+			Kind:       "Ingress",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "default",
+			// not namespaced
+		},
+		Spec: configv1.IngressSpec{
+			Domain: fmt.Sprintf("apps.%s.%s", installConfig.Config.ObjectMeta.Name, installConfig.Config.BaseDomain),
```

Comment:
Is `apps` a standard subdomain for this sort of thing?  As far as I know, we could just use:

```go
fmt.Sprintf("%s.%s", installConfig.Config.ObjectMeta.Name, installConfig.Config.BaseDomain),
```

right?  This is going to be the main entry-point for the cluster's end-users, and I expect folks will want as short a URL as possible for that.  Also, keeping everything one level beneath `{cluster-name}.{base-domain}` means that folks supplying their own X.509 can use a single wildcard cert for more cluster resources (not great for security, but cheaper if you're buying certs instead of using Let's Encrypt or similar).

Reply:
In OCP 3.x, `apps.` has been the conventional prefix, used in most of our example inventories and documentation.  The setting in openshift-ansible is `openshift_master_default_subdomain`, which defaults to "router.default.svc.cluster.local".  Beyond that, I know of nothing that could reasonably be considered a standard.

We have been using the "apps." prefix during 4.0 development, even going as far as to hard-code it in cluster-ingress-operator (pending this change to the installer so that cluster-ingress-operator has something it can use other than a hard-coded value).

@ironcladlou, can you confirm that we want to use this as the default value? Note that as Abhinav noted in an earlier comment on this PR, the default value can be overridden by editing the manifests that the installer produces before creating the cluster.

Reply:
The prefix choice itself ("apps") probably still needs some rehashing, but AFAIK it'll do for now and keep us moving forward.